### PR TITLE
[16.0][MIG] fieldservice_portal: Migration to 16.0

### PR DESCRIPTION
**Pull request by anusriNPS**
MIgrating fieldservice_portal to 16.0 version

### CODE REVIEW

File: fieldservice_portal/views/fsm_order_template.xml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                            >
+                                <!--
+                                careful: this does not work if groupby is not a relation
+                                        towards a model that has a 'name' field
+                                -->
+                                <th colspan="6" class="text-left text-muted">
+                                    <em class="font-weight-normal text-muted">
+                                        Work Orders for
+                                        <t
+                                            t-esc="searchbar_groupby[groupby]['label']"
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: `t-esc` has been deprecated in 15.0 in favour of `t-out`

File: fieldservice_portal/views/fsm_order_template.xml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                                <th>Type</th>
+                                <th>Stage</th>
+                            </tr>
+                        </thead>
+                        <tbody>
+                            <t t-foreach="orders" t-as="order">
+                                <tr>
+                                    <td>
+                                        <a t-attf-href="/my/workorder/#{order.id}">
+                                            <t t-esc="order.name" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/fsm_order_template.xml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                                        </a>
+                                    </td>
+                                    <td>
+                                        <t t-esc="order.description" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/fsm_order_template.xml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                                    </td>
+                                    <td>
+                                        <t
+                                            t-esc="order.sudo().location_id.display_name"
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/fsm_order_template.xml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                                        />
+                                    </td>
+                                    <td>
+                                        <t t-esc="order.type.name" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/fsm_order_template.xml [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                                    </td>
+                                    <td>
+                                        <t t-esc="order.stage_id.name" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/fsm_order_template.xml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                                    )
+                                </t>
+                                <br />
+                                <strong>Country:</strong>
+                                <span t-field="order.location_id.country_id" />
+                                <br />
+                            </div>
+                            <div class="col-md-6">
+                                <strong>Stage:</strong>
+                                <t t-esc="order.stage_id.name" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/fsm_order_template.xml]: also here

File: fieldservice_portal/views/portal_template.xml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                t-attf-class="breadcrumb-item #{'active ' if not fsm_order else ''}"
+            >
+                <a
+                    t-if="fsm_order"
+                    t-attf-href="/my/workorders?{{ keep_query() }}"
+                >Work Orders</a>
+                <t t-else="">Work Orders</t>
+            </li>
+            <li t-if="fsm_order" class="breadcrumb-item active">
+                <t t-esc="fsm_order.name" />
```
Comment (PicchiSeba) [inline fieldservice_portal/views/portal_template.xml]: also here

**Review by PicchiSeba — CHANGES_REQUESTED** [review]
Code review: minor fixes in the views

File: fieldservice_portal/views/fsm_order_template.xml (hunk 1 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 2 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 3 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 4 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 5 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 6 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/fsm_order_template.xml (hunk 7 above)
Comment (anusriNPS) [reply fieldservice_portal/views/fsm_order_template.xml]: Done

File: fieldservice_portal/views/portal_template.xml (hunk 8 above)
Comment (anusriNPS) [reply fieldservice_portal/views/portal_template.xml]: Done

**Comment by anusriNPS** [comment]
Verified views after using t-out instead of t-esc and all are intact

**Comment by PicchiSeba** [comment]
We might need to add some tests, the coverage is quite low.

**Comment by max3903** [comment]
@anusriNPS I don't see Work Orders on the portal:
image

What did I miss? A configuration step that should be added to the readme fragments?

**Comment by anusriNPS** [comment]
@max3903 Work orders are listed while using API /my/workorders, is this API you are looking for or is it required work order to be listed on homepage? Currently, as per available changes, on home page workorders are not displayed
image

**Comment by max3903** [comment]
@anusriNPS My expectation is that the customer can create and track their FSM orders on the portal. So it should appear on the portal page.

This is my opinion. Others can share theirs...

**Comment by max3903** [comment]
Maybe you can document that this is not available and provide the direct path in the readme fragments.

**Comment by brian10048** [comment]
> @anusriNPS My expectation is that the customer can create and track their FSM orders on the portal. So it should appear on the portal page.
> 
> This is my opinion. Others can share theirs...

My expectation is the same. As a customer, you would be able to view your FSM orders from the portal home page just like other documents

File: fieldservice_portal/controllers/fsm_order_portal.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+from odoo.exceptions import AccessError
+from odoo.http import request
+from odoo.osv.expression import OR
+from odoo.tools import groupby as groupbyelem
+
+from odoo.addons.portal.controllers.portal import CustomerPortal, pager as portal_pager
+
+
+class CustomerPortal(CustomerPortal):
+    def _prepare_portal_layout_values(self):
```
Comment (PicchiSeba) [inline fieldservice_portal/controllers/fsm_order_portal.py]: **issue**: missing from Controller

```suggestion

    def _prepare_home_portal_values(self, counters):
        values = super()._prepare_home_portal_values(counters)
        if "fsm_order_count" in counters:
            fsm_order_model = request.env["fsm.order"]
            fsm_order_count = (
                fsm_order_model.search_count([])
                if fsm_order_model.check_access_rights("read", raise_exception=False)
                else 0
            )
            values["fsm_order_count"] = fsm_order_count
        return values

    def _prepare_portal_layout_values(self):
```

File: fieldservice_portal/views/portal_template.xml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        name="Portal My Home : fsm order entries"
+        inherit_id="portal.portal_my_home"
+        priority="45"
+    >
+        <xpath expr="//div[hasclass('o_portal_docs')]" position="inside">
+            <t t-if="fsm_order_count" t-call="portal.portal_docs_entry">
+              <t t-set="title">Work Orders</t>
+              <t t-set="url" t-value="'/my/workorders'" />
+              <t t-set="count" t-value="fsm_order_count" />
+            </t>
```
Comment (PicchiSeba) [inline fieldservice_portal/views/portal_template.xml]: **issue**: missing from portal view

```suggestion
        customize_show="True"
    >
        <xpath expr="//div[hasclass('o_portal_docs')]" position="inside">
            <t t-call="portal.portal_docs_entry">
              <t t-set="title">Work Orders</t>
              <t t-set="url" t-value="'/my/workorders'" />
              <t t-set="placeholder_count" t-value="'fsm_order_count'" />
            </t>
```

**Review by PicchiSeba — CHANGES_REQUESTED** [review]
Code review: following the maintainer suggestions we should add the necessary button to the main portal view

It can be accessed at `/my` with the portal user

File: fieldservice_portal/static/src/js/fsm_order_portal.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+/** @odoo-module */
```
Comment (PicchiSeba) [inline fieldservice_portal/static/src/js/fsm_order_portal.js]: **issue**: pre-commit is whining because the file must be renamed `fsm_order_portal.esm.js`

**Review by PicchiSeba — CHANGES_REQUESTED** [review]

**Review by max3903 — APPROVED** [review]

File: fieldservice_portal/controllers/fsm_order_portal.py (hunk 9 above)
Comment (anusriNPS) [reply fieldservice_portal/controllers/fsm_order_portal.py]: Done

File: fieldservice_portal/views/portal_template.xml (hunk 10 above)
Comment (anusriNPS) [reply fieldservice_portal/views/portal_template.xml]: Done

File: fieldservice_portal/static/src/js/fsm_order_portal.js (hunk 11 above)
Comment (anusriNPS) [reply fieldservice_portal/static/src/js/fsm_order_portal.js]: Updated extension with "esm.js". Thank you @PicchiSeba

**Review by PicchiSeba — APPROVED** [review]
Code review: LGTM

**Review by peluko00 — APPROVED** [review]
LGTM

**Comment by peluko00** [comment]
Seems i'ts ready to merge @max3903

**Pull request merged**